### ISSUE TITLE
game dectection issue

### ISSUE THREAD

**Issue by ahjolinna** [wyyadd/LaLa#25]
I 'm only now started to test if/how the trainers work, and at least with "Baldur's Gate 3", "Hogwarts Legacy"  & "Soul Hackers 2" ... the trainers said that it couldn't detect the game running, my first thought was that I was using [proton-GE](https://github.com/GloriousEggroll/proton-ge-custom/releases) instead of proton-experimental for the games (and you need to be using same version of proton)...but no, that didn't help.

I also tried both [v1.0.7](https://github.com/wyyadd/LaLa/releases/tag/v1.0.7), AppImage and flatpak.

so is this game specific issue, of because I use flatpak version of steam?

**Comment by wyyadd** [comment]
Have you tried firstly launching game then launching trianers?
For custom trainers and non-steam games, LaLa auto-detects them when games are running.

**Comment by ahjolinna** [comment]
there is finally a new flatpak pre-release with nice amount of wayland and sandbox fixes&features (especially if compare to the stable release) : https://github.com/flatpak/flatpak/releases/tag/1.15.6

**Comment by wyyadd** [comment]
> the trainers said that it couldn't detect the game running

Did the error message say "Game not found. For non-Steam games or custom trainers, make sure the game is running"?

> I tested now, at least with Baldur's Gate 3 ...no go ...sadly

Sadly, I can not reproduce it.
LaLa detects running games by using command ```ps aux | grep "SteamLaunch AppId=" | grep -v -i "lala"```

> btw. I was wondering if there is any common (mainstream) game that we share that we could test and see if we have same issues

For me, I've been using Death Stranding for testing because it's a game I play regularly.

This is my AppImage message:
``` bash
leo@leo-MS-7C94:~$ ./LaLa_linux_amd64.AppImage 
(LaLa:13309): Gdk-CRITICAL **: 09:36:53.100: gdk_window_get_state: assertion 'GDK_IS_WINDOW (window)' failed
```

**Comment by ahjolinna** [comment]
well I don't have Death Stranding on Steam, I have it on Epic game store....as it was one of those free games

this is the error:
![image](https://github.com/wyyadd/LaLa/assets/6993072/f607ed04-97a8-4b53-be78-b3d9fdb9896d)


when I launched the game, this time Cities Skylines II and tried the command `ps aux | grep "SteamLaunch AppId=" | grep -v -i "lala"` ...I got this :

```
ahjolin+ 20784  0.0  0.0  29244  5652 ?        S    07:31   0:00 /home/ahjolinna/.var/app/com.valvesoftware.Steam/.local/share/Steam/ubuntu12_32/reaper SteamLaunch AppId=949230 -- /home/ahjolinna/.var/app/com.valvesoftware.Steam/.local/share/Steam/ubuntu12_32/steam-launch-wrapper -- /home/ahjolinna/.var/app/com.valvesoftware.Steam/.local/share/Steam/steamapps/common/SteamLinuxRuntime_sniper/_v2-entry-point --verb=waitforexitandrun -- /home/ahjolinna/.var/app/com.valvesoftware.Steam/.local/share/Steam/compatibilitytools.d/GE-Proton8-23/proton waitforexitandrun /home/ahjolinna/.var/app/com.valvesoftware.Steam/.local/share/Steam/steamapps/common/Cities Skylines II/Launcher/dowser.exe
ahjolin+ 25465  0.0  0.0   6548  2048 pts/0    S+   07:35   0:00 grep --color=auto SteamLaunch AppId=

```

same with Baldurs Gate 3:

```
ahjolin+  7045  0.0  0.0  29244  5652 ?        S    07:49   0:00 /home/ahjolinna/.var/app/com.valvesoftware.Steam/.local/share/Steam/ubuntu12_32/reaper SteamLaunch AppId=1086940 -- /home/ahjolinna/.var/app/com.valvesoftware.Steam/.local/share/Steam/ubuntu12_32/steam-launch-wrapper -- /home/ahjolinna/.var/app/com.valvesoftware.Steam/.local/share/Steam/steamapps/common/SteamLinuxRuntime_sniper/_v2-entry-point --verb=waitforexitandrun -- /home/ahjolinna/.var/app/com.valvesoftware.Steam/.local/share/Steam/compatibilitytools.d/GE-Proton8-23/proton waitforexitandrun /home/ahjolinna/.var/app/com.valvesoftware.Steam/.local/share/Steam/steamapps/common/Baldurs Gate 3/Launcher/LariLauncher.exe
ahjolin+  9576  0.0  0.0   6548  2048 pts/0    S+   07:52   0:00 grep --color=auto SteamLaunch AppId=

```

when I saw that output it mentions only the Launchers exe not the game...so I tested another game that dont have a launcher....and...Octopath Traveler, same error and this command output:
```
ahjolin+ 28118  0.1  0.0  29252  5652 ?        S    08:07   0:00 /home/ahjolinna/.var/app/com.valvesoftware.Steam/.local/share/Steam/ubuntu12_32/reaper SteamLaunch AppId=921570 -- /home/ahjolinna/.var/app/com.valvesoftware.Steam/.local/share/Steam/ubuntu12_32/steam-launch-wrapper -- /home/ahjolinna/.var/app/com.valvesoftware.Steam/.local/share/Steam/steamapps/common/SteamLinuxRuntime_sniper/_v2-entry-point --verb=waitforexitandrun -- /home/ahjolinna/.var/app/com.valvesoftware.Steam/.local/share/Steam/compatibilitytools.d/GE-Proton8-23/proton waitforexitandrun /home/ahjolinna/.var/app/com.valvesoftware.Steam/.local/share/Steam/steamapps/common/OCTOPATH TRAVELER/Octopath_Traveler.exe
ahjolin+ 28668  0.0  0.0   6548  2048 pts/0    S+   08:07   0:00 grep --color=auto SteamLaunch AppId=

```
and thats wrong exe

same with Soul Hackers 2:

```
ahjolin+ 31241  0.1  0.0  29252  5652 ?        S    08:10   0:00 /home/ahjolinna/.var/app/com.valvesoftware.Steam/.local/share/Steam/ubuntu12_32/reaper SteamLaunch AppId=1777620 -- /home/ahjolinna/.var/app/com.valvesoftware.Steam/.local/share/Steam/ubuntu12_32/steam-launch-wrapper -- /home/ahjolinna/.var/app/com.valvesoftware.Steam/.local/share/Steam/steamapps/common/SteamLinuxRuntime_sniper/_v2-entry-point --verb=waitforexitandrun -- /home/ahjolinna/.var/app/com.valvesoftware.Steam/.local/share/Steam/compatibilitytools.d/GE-Proton8-23/proton waitforexitandrun /home/ahjolinna/.var/app/com.valvesoftware.Steam/.local/share/Steam/steamapps/common/SOUL HACKERS2/SOUL HACKERS2.exe
ahjolin+ 31935  0.0  0.0   6548  2048 pts/0    S+   08:10   0:00 grep --color=auto SteamLaunch AppId=

```

---------------

my system:


```
Operating System: openSUSE MicroOS 20231113
KDE Plasma Version: 5.27.9
KDE Frameworks Version: 5.112.0
Qt Version: 5.15.11
Kernel Version: 6.6.1-1-default (64-bit)
Graphics Platform: Wayland
Processors: 16 × AMD Ryzen 7 5700G with Radeon Graphics
Memory: 46.8 GiB of RAM
Graphics Processor: NVIDIA GeForce RTX 3060 Ti/PCIe/SSE2
GPU Driver Version : 545.29.02
```

**Comment by wyyadd** [comment]
> Game is not detected, please launch the game before activating cheats.

This error messge is sent by trainers. I guess **the game process name** in the trainers page must be flickering.
Are you sure the game and trainer use same proton and winepfx?
Your steam path should be ```~/.var/app/com.valvesoftware.Steam/.local/share/Steam```

**Comment by ahjolinna** [comment]
>This error messge is sent by trainers. I guess the game process name in the trainers page must be flickering.

yes the game process names are flickering, but also that command seems to find launcher exe instead of the games exe file

> Are you sure the game and trainer use same proton and winepfx?

well [proton-ge](https://github.com/GloriousEggroll/proton-ge-custom) location is at `~/.var/app/com.valvesoftware.Steam/data/Steam/compatibilitytools.d/`

but even when I use proton-experimental (or v8.0) I get same error ..both exist at `~/.var/app/com.valvesoftware.Steam/data/Steam/steamapps/common/*` and the trainer should be correct 

> BTW: I also use Hollow Knight as a test on steam.

well I dont have Hollow Knight either :D

**Comment by wyyadd** [comment]
Here is how LaLa run trainers:
``` bash
STEAM_COMPAT_CLIENT_INSTALL_PATH=~/.var/app/com.valvesoftware.Steam/.local/share/Steam \
STEAM_COMPAT_DATA_PATH=~/.var/app/com.valvesoftware.Steam/.local/share/Steam/steamapps/compatdata/{APPID} \
~/.var/app/com.valvesoftware.Steam/.local/share/Steam/steamapps/common/{Proton Version}/proton run path/to/trainers.exe
```
Can you try this command?
You can download trianers at [fling](https://flingtrainer.com/all-trainers-a-z/)

**Comment by ahjolinna** [comment]
well I tried with [Soul Hackers 2](https://flingtrainer.com/trainer/soul-hackers-2-trainer/), the process name was still flickering 
![image](https://github.com/wyyadd/LaLa/assets/6993072/27c79093-f7cd-43ad-81ed-41e94e4221fa)
```
STEAM_COMPAT_CLIENT_INSTALL_PATH=/home/ahjolinna/.var/app/com.valvesoftware.Steam/.local/share/Steam STEAM_COMPAT_DATA_PATH=/home/ahjolinna/.var/app/com.valvesoftware.Steam/.local/share/Steam/steamapps/compatdata/1777620 "/home/ahjolinna/.var/app/com.valvesoftware.Steam/data/Steam/steamapps/common/Proton - Experimental/proton" run "/home/ahjolinna/Downloads/Soul Hackers 2 v1.01 Plus 25 Trainer.exe"
fsync: up and running.
wine: RLIMIT_NICE is <= 20, unable to use setpriority safely
Unable to read VR Path Registry from C:\users\steamuser\AppData\Local\openvr\openvrpaths.vrpath
```

when I tried

```
STEAM_COMPAT_CLIENT_INSTALL_PATH=~/.var/app/com.valvesoftware.Steam/.local/share/Steam STEAM_COMPAT_DATA_PATH=~/.var/app/com.valvesoftware.Steam/.local/share/Steam/steamapps/compatdata/1777620 "~/.var/app/com.valvesoftware.Steam/data/Steam/steamapps/common/Proton - Experimental/proton" run "~/Downloads/Soul Hackers 2 v1.01 Plus 25 Trainer.exe"
bash: ~/.var/app/com.valvesoftware.Steam/data/Steam/steamapps/common/Proton - Experimental/proton: No such file or directory

```

I have currently the flu so if I miss something basic its because of that

**Comment by wyyadd** [comment]
I think I figure out the reason. 
It is because you use flatpak version of steam.

Below is my tests:
normal Steam --- normal lala -- work
normal Steam - flatpak lala -- work 
flatpak Steam --- normal lala -- not work
flatpak Steam --- flatpak lala -- not work

**Comment by ahjolinna** [comment]
well that will sucks especially for immutable distros that uses flatpak version of steam.
and as (open)SUSE and fedora/RHL (and ubuntu, but will use their own snap thing) are going towards this immutable ...this is little bit annoying issue.

SteamOS is immutable but doesn't current at least use flatpak version, what I understand the flatpak version of steam client is not even officially supported (community build)

**Comment by ahjolinna** [comment]
okay I can confirm that this is an flatpak issue, as I just installed the normal steam (and all the dependencies with it) on my system ...sigh.

so question is that is there an workaround? other than installing the non-flatpak version of steam (which isn't an option at least on Fedora/RHEL implentation of immutable OS*)


------------

*SUSE's implementation is based on btrfs cow capabilities limiting it to the btrfs' limits for example in RAID support, whereas since Fedora/RHEL doesn't support btrfs they had to implement transactioness on the package manager level making it more complicated but compatible with all filesystems.


You can read more about it here https://www.theregister.com/2023/02/16/bulletproof_linux/

I have no idea how does Fedora/RHEL handle configuration changes in /etc when rebasing, and how Ubuntu has done theirs in general (other than your have to use snaps instead of flatpak).

SteamOS approach is close SUSE's, aka is btrfs based but other details are different

**Comment by wyyadd** [comment]
I'm currently looking for a workaround. I've noticed that SteamTinkerLaunch supports Flatpak Steam, and they have a [wiki](https://github.com/sonic2kk/steamtinkerlaunch/wiki/Steam-Flatpak) that might provide insights. 
However, I haven't found the specific steps yet. I need to dig deeper to understand the process better.

**Comment by ahjolinna** [comment]
I found a mod utility for Bethesda games that has flatpak version ...yes its way different than yours (for example its made in Qt) but maybe there is some help how it works (especially  on flatpak) https://flathub.org/apps/io.github.loot.loot


they do have this issue open but some interesting info https://github.com/loot/loot/issues/1900

**Comment by wyyadd** [comment]
Thank you for providing the information. I will take a look :)
Recently, I've been playing the Stardew Valley, and I also found a similar issue on its [wiki](https://stardewvalleywiki.com/Modding:Installing_SMAPI_on_Linux). 
> instructions for Flatpak 
> When using Flatpak, you need to change the default game location: 

Although I changed the default game location, LaLa still can't detect games. :(
P.S.: Sorry for not updating recently. A little busy preparing to study in Canada in the coming days :D

**Comment by ahjolinna** [comment]
> Thank you for providing the information. I will take a look :) Recently, I've been playing the Stardew Valley, and I also found a similar issue on its [wiki](https://stardewvalleywiki.com/Modding:Installing_SMAPI_on_Linux).
> 
> > instructions for Flatpak
> > When using Flatpak, you need to change the default game location:
> 

good that you also found some tips how to handle the flatpak issue, hopefully it can be resolved with just default location change (then again like wayland, flatpak isn't yet in mainstream usage so when it become more used by normal users there will be more changes/fixes)

> Although I changed the default game location, LaLa still can't detect games. :( P.S.: Sorry for not updating recently. A little busy preparing to study in Canada in the coming days :D

no worry we are all busy at times, I'm currently preparing for upcoming elections this year here in Finland as I'm electoral officer ( a nice side work, I get ~700€ from one days work...well minus taxes)

**Comment by wyyadd** [comment]
I think have fixed this issue at ![Pre-release 1.0.9](https://github.com/wyyadd/LaLa/releases/tag/v1.0.9)
To use LaLa in flatpak steam, you should:
1. Add LaLa as non-steam game.
2. Add environment variable **IN_FLATPAK_STEAM=1** to LaLa's Launch Options.

What I did:
Flatpak steam and games run in a **sandbox**.   
If LaLa is launched as a non-steam game, it will in the same sandbox with steam, which means LaLa can detect games and modify their memory.

In my testing, I added AppImage LaLa as a non-steam game. I haven't tried Flatpak LaLa. However, I assume that flatpak lala may not work because flatpak lala and steam runs in different sanboxes. I would appreciate it if you could try add flatpak lala as a non-steam game and do some tests. :)

My testing environment is so horrible (I don't have a amd64 PC at present, so I have to use my arm Macbook to emulate amd64 Linux. The speed is so SLOW).

**Comment by wyyadd** [comment]
I also updated the readme.

For user who uses [Flatpak Steam](https://flathub.org/apps/com.valvesoftware.Steam), you also need to:
- Add AppImage LaLa as a Non-Steam game.
- Use ```IN_FLATPAK_STEAM=1 %command% --appimage-extract-and-run``` as LaLa's launch options.
- Open LaLa through steam.
